shared/p2p/service.go

```diff
@@ -91,6 +77,85 @@ func (s *Server) Stop() error {
 	return nil
 }
 
+// RegisterTopic with a message and the adapter stack for the given topic. The
+// message type provided will be feed selector for emitting messages received
+// on a given topic.
+//
+// The topics can originate from multiple sources. In other words, messages on
+// TopicA may come from direct peer communication or a pub/sub channel.
+func (s *Server) RegisterTopic(topic string, message interface{}, adapters ...Adapter) {
+	msgType := reflect.TypeOf(message)
+	log.WithFields(logrus.Fields{
+		"topic": topic,
+	}).Debug("Subscribing to topic")
+
+	s.topicMapping[msgType] = topic
+
+	sub, err := s.gsub.Subscribe(topic)
+	if err != nil {
+		log.Errorf("Failed to subscribe to topic: %v", err)
+		return
+	}
+	feed := s.Feed(msgType)
+
+	go (func() {
+		defer sub.Cancel()
+		for {
+			msg, err := sub.Next(s.ctx)
+
+			if s.ctx.Err() != nil {
+				log.WithError(s.ctx.Err()).Debug("Context error")
+				return
+			}
+
+			if err != nil {
+				log.Errorf("Failed to get next message: %v", err)
+				return
+			}
+
+			// Reverse adapter order
+			for i := len(adapters)/2 - 1; i >= 0; i-- {
+				opp := len(adapters) - 1 - i
+				adapters[i], adapters[opp] = adapters[opp], adapters[i]
+			}
+
+			var h Handler = func(ctx context.Context, pMsg Message) {
+				s.emit(feed, msg, msgType)
+			}
+
+			pMsg := Message{}
+
+			for _, adapter := range adapters {
+				h = adapter(h)
+			}
+
+			ctx, _ := context.WithTimeout(s.ctx, 10*time.Second)
+			h(ctx, pMsg)
+		}
+	})()
+
+}
+
+func (s *Server) emit(feed *event.Feed, msg *floodsub.Message, msgType reflect.Type) {
+	// TODO: reflect.Value.Interface() can panic so we should capture that
+	// panic so the server doesn't crash.
+	d, ok := reflect.New(msgType).Interface().(proto.Message)
+	if !ok {
+		log.Error("Received message is not a protobuf message")
+		return
+	}
+	if err := proto.Unmarshal(msg.Data, d); err != nil {
+		log.Errorf("Failed to decode data: %v", err)
+		return
+	}
+
+	i := feed.Send(Message{Data: d})
```

Comment:
Here's the definition for Message:
```
type Message struct {
	// Peer represents the sender of the message.
	Peer Peer
	// Data can be any type of message found in sharding/p2p/proto package.
	Data interface{}
}
```

Can type `interface{}` type for `Data` be replaced with `proto.Message`? Allowing any type here seems unnecessarily lax, especially considering that the value was cast down to `proto.Message` at the beginning of this function.

Reply:
Sounds good to me